HiFi-GAN vocoder for discrete tokens

Chaanks opened the pull request:
## What does this PR do?

This PR adds support for HiFi-GAN to work with new SSL Discrete Tokens and supports for bitrate-scalable training.
Including recipe for LJSpeech and LibriTTS.

Imade a few changes to the LibriTTS data preparation script. Please check them.

<details>
  <summary><b>Before submitting</b></summary>

- [x ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [x] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [x] Review the self-review checklist to ensure the code is ready for review

</details>

<!--

🎩 Magic happens when you code. Keep the spells flowing!

-->

poonehmousavi commented:
Thanks @Chaanks  for this PR.. It looks really good. I just added couple of comments. Apart from them:

1. make sure that all the test are working: You could use following command to run the  test: 
`python -c 'from tests.utils.recipe_tests import run_recipe_tests; print("TEST FAILED!") if not(run_recipe_tests(filters_fields=["Task"], filters=[["TTS"]], do_checks=True, run_opts="--device=cuda")) else print("TEST PASSED")'`

2. Please transfer  all the vocoder checkpoints from your HF to SPeechbrain and fix the reference in the readme and docstring.

Chaanks commented on recipes/LJSpeech/TTS/vocoder/hifi_gan_unit/train.py:
```diff
@@ -423,10 +416,20 @@ def audio_pipeline(utt_id, wav, segment):
+        offsets = np.arange(num_layer) * hparams["num_clusters"]
+        code = code + offsets + 1
+
+        if hparams["layer_drop"]:
```
I think it's better to do it at the utterance level rather than the batch level

Chaanks commented on recipes/LJSpeech/TTS/vocoder/hifi_gan_unit/extract_code.py:
```diff
@@ -9,21 +9,33 @@
 VALID_JSON = "valid.json"
 TEST_JSON = "test.json"
 
+ENCODER_CLASSES = {
```
"I prefer the way I'm doing it currently. The ssl_model is only used for token extraction and not during training, so I don't see why we need to instantiate the model in the YAML and it will require manual cleanup to free memory for HiFi-GAN training.

poonehmousavi replied:
i dont see what is the problem in defining it in yaml file instead of the python file... why ther isa difference for manual cleanup?

Chaanks commented on recipes/LJSpeech/TTS/vocoder/hifi_gan_unit/extract_code.py:
```diff
@@ -169,26 +180,33 @@ def extract_ljspeech(
-    logger.info(f"Loading K-means model from {kmeans_ckpt} ...")
-    kmeans_model = joblib.load(open(kmeans_ckpt, "rb"))
-    kmeans_model.verbose = False
+    discrete_encoder = DiscreteSSL(
```
See my other comment. I added a parameter for the save path.

Chaanks commented on recipes/LJSpeech/TTS/vocoder/hifi_gan_unit/extract_code.py:
```diff
@@ -169,26 +180,33 @@ def extract_ljspeech(
+        raise TypeError("Not a supported Encoder")
+
+    encoder_class = ENCODER_CLASSES[encoder_type]
+    encoder = encoder_class(
```
See my other comment.

mravanelli commented:
Thank you @Chaanks! Here are some comments:

1. hifi_gan_unit => Is this name from the literature?

2. encoder_hub: facebook/hubert-large-ll60k => Add comments in the YAML and in the README file with the options that users need to set for wav2vec2 and wavLM.

3. In the README.md: This vocoder is a neural network designed to transform discrete self-supervised representations into waveform data and is suitable for speech-to-speech translation on top of CVSS/S2ST models  => This is not only needed for speech translation but in general for generative tasks like TTS, speech enhancement, separation, voice cloning, etc. We should clarify that. Maybe we can also mention that we use this in our DASB benchmark

4. In the yaml, we need to add a comment on "kmeans_folder:" to describe what it is. Is this supposed to be a !PLACEHOLDER variable?

5. In https://huggingface.co/speechbrain/hifigan-hubert-l1-3-7-12-18-23-k1000-LibriTTS, we can add the reference to our interspeech paper (or any other paper) that describes well the scalable vocoder. We need to reference that also in README.md

6. In README.md, we normally reference the Dropbox folder with the checkpoints and logs. Do we have it?

7. It looks like recipes/LJSpeech/TTS/quantization is outdated, right? If so, it can be deleted and the README file should be updated with the right instructions.

I would recommend to double check if all the vocoder training is running when following step by step the instructions in the README file.

Chaanks commented:
1. Yes it comes from this work: [Speech Resynthesis from Discrete Disentangled Self-Supervised Representations](https://arxiv.org/abs/2104.00355) and was then used for S2ST. What do you think of renaming it `hifi_gan_discrete` ? It will better correlate with the discrete module of SB.

6. Yes, we have it uploaded for the Benchmark paper. Should we directly add a reference to this?

7.  Deleted. I also added backward compatibility to the Vocoder for previous models without scalability (https://huggingface.co/speechbrain/tts-hifigan-unit-hubert-l6-k100-ljspeech).

poonehmousavi commented:
> 1. Yes it comes from this work: [Speech Resynthesis from Discrete Disentangled Self-Supervised Representations](https://arxiv.org/abs/2104.00355) and was then used for S2ST. What do you think of renaming it `hifi_gan_discrete` ? It will better correlate with the discrete module of SB.
> 2. Yes, we have it uploaded for the Benchmark paper. Should we directly add a reference to this?
> 3. Deleted. I also added backward compatibility to the Vocoder for previous models without scalability (https://huggingface.co/speechbrain/tts-hifigan-unit-hubert-l6-k100-ljspeech).

1.I personally prefer the name hifi_gan_discrete.
2. I think we could add the interspeech paper as a reference

mravanelli commented:
Thank you @Chaanks!
It looks like everything is running. 
I only have this final comment on the README files (`speechbrain/recipes/LJSpeech/TTS/README.md` and `speechbrain/recipes/LJSpeech/TTS/README.md`).

In the README.md it is written, "The kmeans_folder, kmeans_dataset and num_clusters should be specified based on SSL_Quantization."

This part should be elaborated more. Here, to simplify running the recipe, we created an HF repository with all the needed models.

However, it could be important to also describe what users should do to use their own quantization model or to just replicate what we did from scratch. I think they should go to speechbrain/recipes/LJSpeech/quantization/, run the quantizers, and store it somewhere. These steps should be described and tested as I feel we need to better connect in the documentation the vocoder part and the quantization parts.

mravanelli commented:
Thank you @Chaanks!
